mlflow/deployments/base.py

```diff
@@ -87,7 +87,81 @@ def __init__(self, target_uri):
+                     name, raises a :py:class:`mlflow.exceptions.MlflowException`.
+        :param config: (optional) Dict containing target-specific configuration for the
+                       endpoint.
+        :return: Dict corresponding to created endpoint, which must contain the 'name' key.
+        """
+        pass
+
+    @abc.abstractmethod
+    def update_endpoint(self, name, config=None):
+        """
+        Update the endpoint with the specified name. You can update any target-specific attributes
+        of the endpoint (via `config`). By default, this method should block until the update
+        completes (i.e. until it's possible to create a deployment within the endpoint). See
+        target-specific plugin documentation for additional detail on support for asynchronous
+        update and other configuration.
+
+        :param name: Unique name of endpoint to update
+        :param config: (optional) dict containing target-specific configuration for the
+                       endpoint
+        :return: None
+        """
+        pass
+
+    @abc.abstractmethod
+    def delete_endpoint(self, name):
+        """
+        Delete the endpoint with name ``name`` from the specified target. Deletion should be
+        idempotent (i.e. deletion should not fail if retried on a non-existent deployment).
+
+        :param name: Name of endpoint to delete
+        :return: None
+        """
+        pass
+
+    @abc.abstractmethod
+    def list_endpoints(self):
+        """
+        List endpoints in the specified target. This method is expected to return an
+        unpaginated list of all endpoints (an alternative would be to return a dict with
+        an 'endpoints' field containing the actual endpoints, with plugins able to specify
+        other fields, e.g. a next_page_token field, in the returned dictionary for pagination,
+        and to accept a `pagination_args` argument to this method for passing
+        pagination-related args).
+
+        :return: A list of dicts corresponding to endpoints. Each dict is guaranteed to
+                 contain a 'name' key containing the endpoint name. The other fields of
+                 the returned dictionary and their types may vary across targets.
+        """
+        pass
+
+    @abc.abstractmethod
+    def get_endpoint(self, name):
+        """
+        Returns a dictionary describing the specified endpoint, throwing a
+        py:class:`mlflow.exception.MlflowException` if no endpoint exists with the provided
+        name.
+        The dict is guaranteed to contain an 'name' key containing the endpoint name.
+        The other fields of the returned dictionary and their types may vary across targets.
+
+        :param name: Name of endpoint to fetch
+        """
+        pass
+
+    @abc.abstractmethod
+    def create_deployment(self, name, model_uri, flavor=None, config=None, endpoint_uri=None):
```

Comment:
Could `endpoint_uri` be folded into `config`? It decouples the concepts of deployment from endpoint. For use cases w/o endpoint, a user doesn't have to worry about it at all.

Ditto below for other *_deployment() functions.

Reply:
@trangevi @jinzhang21 As we discussed offline, I think that deployments and endpoints are strongly coupled on certain platforms (e.g. AzureML) and pretty strongly linked on other platforms (SageMaker, TorchServe). Accordingly, providing users with a single function for creating a deployment and adding it to an endpoint seems like a good idea because it's broadly applicable. If we're having trouble figuring out how to refer to an endpoint (i.e. should it be a string name? should it be a URI? should the URI embed hostname information?), perhaps we can call this parameter `endpoint` for deployment-related methods (e.g. `create_deployment`). This provides us the flexibility to introduce references to endpoints by name, URI, and / or something else in the future. We could initially start with supporting names (this is consistent with the CRUD methods for endpoints introduced in this PR). Thoughts?

Reply:
Naming it as `endpoint` SGTM.

Reply:
Switched to endpoint, as discussed